[BotService] add "bot directline site" command group

stevengum opened the pull request:
___

**Description**  
Add `az bot directline site` command group to manage sub resources

Indirectly 

**Testing Guide**  
`az bot directline site create` takes an existing Bot with a Direct Line Channel Registration and adds a DirectLineSite. The required parameters are:
 - resource_group_name
 - resource_name (name of the existing Bot)
 - site_name (name of the DirectLineSite)

`az bot directline site list` takes an existing Bot with a Direct Line Channel Registration lists all registered sites. The required parameters are:
 - resource_group_name
 - resource_name (name of the existing Bot)

`az bot directline site update` updates an existing Direct Line Channel site. The required parameters are:
 - resource_group_name
 - resource_name (name of the existing Bot)
 - site_name (name of the DirectLineSite)

`az bot directline site delete` deletes an existing Direct Line Channel site. If the site does not exist, no error is thrown. The required parameters are:
 - resource_group_name
 - resource_name (name of the existing Bot)
 - site_name (name of the DirectLineSite)

**History Notes**  
<!--If your PR is not customer-facing, use {Component Name} in the PR title. Otherwise, use [Component Name] to allow our pipeline to add the title as a history note. If you need multiple history notes or would like to overwrite the note from the PR title, please fill in the following templates.-->

[BotService] az bot directline site: Add new command group for Direct Line site management. The supported commands are `create`, `delete`, `list` and `update`.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

yonzhan commented:
BotService

haroldrandom commented on src/azure-cli/azure/cli/command_modules/botservice/_params.py:
```diff
@@ -166,6 +166,29 @@ def load_arguments(self, _):
 
+    with self.argument_context('bot directline site create') as c:
+        c.argument('enable_enhanced_auth', help='If true, enables enhanced authentication features. Must be true for --trusted-origins parameters to work.', arg_type=get_three_state_flag())
+        c.argument('is_enabled', options_list=['--is-enabled'], arg_type=get_three_state_flag(), help='Add the Direct Line Site in an enabled state.')
```
```suggestion
        c.argument('is_enabled', options_list=['--enabled'], arg_type=get_three_state_flag(), help='Add the Direct Line Site in an enabled state.')
```
I guess `enabled` should be enough to represent your indent to user.

stevengum replied:
Good catch! Done.

haroldrandom commented on src/azure-cli/azure/cli/command_modules/botservice/commands.py:
```diff
@@ -80,5 +86,11 @@ def load_command_table(self, _):
+        g.command('create', 'create')
+        g.command('delete', 'delete')
+        g.command('list', 'list')
+        g.command('update', 'update')
```
generic_update_command ?

stevengum replied:
I didn’t the generic update command because the site is a complex sub-resource with no direct API.

The update API called here is for all Channels (Direct Line, MS Teams, Slack, etc). The Direct Line channel’s properties contains a list of sites, so to preserve any pre-existing sites the module needs to fetch the channel and store them for the lifetime of the command.

stevengum commented:
CI Style check blocked on: https://github.com/Azure/azure-cli/pull/13493

qwordy commented on src/azure-cli/azure/cli/command_modules/botservice/_params.py:
```diff
@@ -166,6 +166,29 @@ def load_arguments(self, _):
+        c.argument('enable_enhanced_auth', help='If true, enables enhanced authentication features. Must be true for --trusted-origins parameters to work.', arg_type=get_three_state_flag())
+        c.argument('is_enabled', options_list=['--enabled'], arg_type=get_three_state_flag(), help='Add the Direct Line Site in an enabled state.')
+        c.argument('support_v3', help='If true, v3 protocol will be enabled on the channel.', arg_type=get_three_state_flag())
+        c.argument('show_secrets', options_list=['--with-secrets'], help='Show secrets in response for the channel.', arg_type=get_three_state_flag())
```
`action='store_true'` instead of `arg_type=get_three_state_flag()`

stevengum replied:
Thanks @qwordy! I'll get this change in.

stevengum replied:
Should I plumb this through for the new parameters that use `arg_type=get_three_state_flag()`?

Juliehzl commented on src/azure-cli/azure/cli/command_modules/botservice/_params.py:
```diff
@@ -166,6 +166,29 @@ def load_arguments(self, _):
         c.argument('trusted_origins', nargs='+', help='Space separated Trusted Origins URLs (must use HTTPS) e.g. --trusted-origins https://mybotsite1.azurewebsites.net https://mybotsite2.azurewebsites.net')
 
+    with self.argument_context('bot directline site create') as c:
+        c.argument('enable_enhanced_auth', help='If true, enables enhanced authentication features. Must be true for --trusted-origins parameters to work.', arg_type=get_three_state_flag())
+        c.argument('is_enabled', options_list=['--enabled'], arg_type=get_three_state_flag(), help='Add the Direct Line Site in an enabled state.')
```
These two arguments are confused to me. They are two features to enable, right?

stevengum replied:
They are for separate properties/features... But your feedback makes me think I should remove `is_enabled` as it seems more idiomatic that someone creating a resource in this context would want it enabled. From this perspective `is_enabled` is just unnecessary noise.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/botservice/_params.py:
```diff
@@ -166,6 +166,29 @@ def load_arguments(self, _):
+    with self.argument_context('bot directline site create') as c:
+        c.argument('enable_enhanced_auth', help='If true, enables enhanced authentication features. Must be true for --trusted-origins parameters to work.', arg_type=get_three_state_flag())
+        c.argument('is_enabled', options_list=['--enabled'], arg_type=get_three_state_flag(), help='Add the Direct Line Site in an enabled state.')
+        c.argument('support_v3', help='If true, v3 protocol will be enabled on the channel.', arg_type=get_three_state_flag())
```
Will we have other protocols to support?

stevengum replied:
At the moment, I don't believe so... But if we did, we would definitely evaluate moving away from separate flags for each protocol and creating a better UX for indicating protocol support.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/botservice/_params.py:
```diff
@@ -166,6 +166,29 @@ def load_arguments(self, _):
+        c.argument('trusted_origins', nargs='+', help='Space separated Trusted Origins URLs (must use HTTPS) e.g. --trusted-origins https://mybotsite1.azurewebsites.net https://mybotsite2.azurewebsites.net')
+
+    with self.argument_context('bot directline site delete') as c:
+        c.argument('site_name', options_list=['-s', '--site-name'], help='Name of the Directline channel site.')
```
Can we consider abstracting site_name to a name type using CLIArgumentType? What about ['-n', '--name'] for options_list?

stevengum replied:
Thanks for the comment! We require the name of the bot that owns the directline site sub-resource, which we accept via `['-n', '--name']`. We use these options throughout the command group (az bot).

With that in mind, I think this should continue with `['-s', '--site-name']`, but if there are other suggestions for the parameter I'm happy to discuss it further  😄

Juliehzl commented on src/azure-cli/azure/cli/command_modules/botservice/_params.py:
```diff
@@ -166,6 +166,29 @@ def load_arguments(self, _):
+        c.argument('show_secrets', options_list=['--with-secrets'], help='Show secrets in response for the channel.', arg_type=get_three_state_flag())
+
+    with self.argument_context('bot directline site update') as c:
+        c.argument('enable_enhanced_auth', help='If true, enables enhanced authentication features. Must be true for --trusted-origins parameters to work.', arg_type=get_three_state_flag())
```
It seems it is duplicated with `create` command. I will prefer to make them in one place.

stevengum replied:
@Juliehzl, thanks for the comment! I could keep it in the update command if that is preferable, my intent was to reduce user steps/commands to provision a DirectLineChannelSite, but still have the property configurable in the end.

Should I remove this parameter from `bot directline site create`?

stevengum commented:
@qwordy, @Juliehzl, @haroldrandom sorry for the delay in addressing feedback! My team is currently planning our next milestones, so I will push to have this PR done ASAP this week.

yungezz commented:
hi @stevengum any update on this PR?

qwordy commented:
hi @stevengum, do you still plan to resume this PR? Is it abandoned?

stevengum commented:
@qwordy sorry for the delay, this PR will be reimplemented by another BF team member. Closing PR.
